xtask/src/ui.rs

```diff
@@ -0,0 +1,480 @@
+        .arg("--manifest-path=./workspace/Cargo.toml")
+        .utf8_output(msg_info)?
+        .compiles()?;
+
+    // check within a workspace
+    let workspace_dir = path.join("workspace");
+    cross_command("build", target, image, &workspace_dir, None)?
+        .utf8_output(msg_info)?
+        .compiles()?;
+    cross_command("run", target, image, &workspace_dir, None)?
+        .args(["--features", "dependencies"])
+        .utf8_output(msg_info)?
+        .runs("Hello from test-workspace, bin/dependencies.rs")?;
+    check_file_exists(
+        &workspace_dir,
+        &format!("target/{target}/debug/dependencies"),
+    )?;
+
+    // check we can run custom binaries
+    cross_command("run", target, image, &workspace_dir.join("binary"), None)?
+        .utf8_output(msg_info)?
+        .runs("Hello from binary, binary/src/main.rs")?;
+    check_file_exists(&workspace_dir, &format!("target/{target}/debug/binary"))?;
+
+    // check using a custom target directory
+    cross_command("build", target, image, &workspace_dir, None)?
+        .args(["--features", "dependencies"])
+        .args(["--target-dir", "custom"])
+        .utf8_output(msg_info)?
+        .compiles()?;
+    check_file_exists(
+        &workspace_dir,
+        &format!("custom/{target}/debug/dependencies"),
+    )?;
+
+    // check using a fake toolchain
+    let fake_toolchain = "cross-fake-toolchain";
+    cross_command("build", target, image, &workspace_dir, Some(fake_toolchain))?
+        .utf8_output(msg_info)?
+        .no_toolchain(fake_toolchain)?;
+
+    // test with an invalid container engine
+    cross_command("build", target, image, &workspace_dir, None)?
+        .env("CROSS_CONTAINER_ENGINE", "cross-fake-engine")
+        .utf8_output(msg_info)?
+        .no_engine()?;
+
+    // test with an invalid image
+    let fake_image = "ghcr.io/cross-rs/fake-image:latest";
+    cross_command("build", target, fake_image, &workspace_dir, None)?
+        .utf8_output(msg_info)?
+        .no_image(fake_image)?;
+
+    // TODO(ahuszagh)
+    //  Test other missing config options
+    //      Ex: invalid platforms
+
+    Ok(())
+}
+
+// TODO(ahuszagh) Needs msg_info
+fn cross_util_ui() -> cross::Result<()> {
+    // TODO(ahuszagh)
+    //  Test help
+    //  Test version
+    //  Test images, volumes, containers, clean
+
+    Ok(())
+}
+
+// Adapted from
+// https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/tests/testsuite/support/mod.rs#L507
+fn target_dir() -> PathBuf {
+    env::current_exe()
+        .ok()
+        .map(|mut path| {
+            path.pop();
+            if path.ends_with("deps") {
+                path.pop();
+            }
+            path
+        })
+        .unwrap()
+}
+
+// Adapted from
+// https://github.com/assert-rs/assert_cmd/blob/363daac637b5f20ba3524ccf648de5f96f0dc45f/src/cargo.rs#L198-L208
+fn cargo_bin(bin: impl AsRef<str>, cwd: impl AsRef<Path>) -> cross::Result<Command> {
```

Comment:
This was the only item we needed from `assert_cmd`, which is very easy to implement itself. It's not worth dragging a large dependency, which prevents us from using `run_and_get_output` with the printed commands, and poorer debug messages, for ~30 lines of code.